Updated ZNG/BZNG samples and script to check their md5sums

philrz opened the pull request:
Up to now the sample data repo has contained Zeek logs in default TSV and NDJSON formats, as well as a set in ZSON format as output from `zq`. The ZSON format has changed significantly since they were first generated (nested records support, to name one) and BZSON output was also added.

This PR includes updated new ZSON/BZSON outputs based on `zq` commit `6848d9e`. I've also added a set of MD5 hashes for each ZSON/BZSON log and a `check_md5sums.sh` script that can can be used to check if future `zq` versions generate the same output. In addition to being helpful for catching bugs, my hope is: 

1.  This will allow us keep the ZSON/BZSON samples up to date as a reference data set
2. This will give me a heads-up as to whether any `zq` docs changes need to be made in reaction to intentional changes in the ZSON/BZSON data formats

The name changes to ZNG/BZNG are still pending, so I'll need to push some renames in this PR before merging. However, I wanted to get it up so the approach/script could be reviewed sooner, and I may be able to start working off the updated sample data in docs updates.

Once this is committed, it's my hope someone may help me rig up the automation to run the check script automatically when PRs merge to `zq` so we have an early warning on the output changes that'll need to ripple through here and on to the docs.

Commit:
Updated ZSON/BZSON samples and script to check their md5sums

Commit:
Remove scratch file

Commit:
Clarify script output after a diff

henridf commented on scripts/check_md5sums.sh:
```diff
@@ -0,0 +1,34 @@
+set -e
+
+if [ $# -ne 1 ] || ! { [ "$1" = "zson" ] || [ "$1" = "bzson" ]; }; then
+  echo 'Must specify output type to be cheecked: "zson" or "bzson"'
```
s/cheecked/checked/

nit: i think we tend to say "output format" more than "output type"

mikesbrown replied:
Just pointing out there is also "log format" above.

philrz replied:
Ok, I fixed the typo, and opted for "output format".

henridf commented on scripts/check_md5sums.sh:
```diff
@@ -0,0 +1,34 @@
+done
+
+echo
+echo "diff'ing current \"zq -f $ZNG_TYPE\" output hashes vs. committed hashes:"
```
It would be nice to compare file by file rather than hashing everything first and then comparing. but that's a feature request rather than a script change request :)

philrz replied:
Yeah, fair enough. I'm figuring I'll start simple and see what we find before adding more features. Thankfully the data set is so small that it can do the full run through of the hashes in just a few seconds.

Not sure if you were maybe thinking of actually walking the user through the _diffs_ file-by-file. That's something I thought of as well, but decided against at this stage.

henridf replied:
Yes that was the motivation. But as I said, feature request!

mikesbrown commented on README.md:
````diff
@@ -28,4 +29,44 @@ The data set was made from the several PCAP files in the [2018](https://archive.
+
+```
+# for file in zeek-default/*; do gzcat $file | zq -f zson - | gzip > zson/$(basename $file | sed 's/\.log\.gz//').zson.gz ; done
+# for file in zeek-default/*; do gzcat $file | zq -f bzson - | gzip > bzson/$(basename $file | sed 's/\.log\.gz//').bzson.gz ; done
````
Nit1. Use backslashes and other constructions to trim line length. At https://github.com/mccanne/zq-sample-data/tree/zng-redo#creation you can see this needing scroll bars.
Nit2: shellcheck your example bash ;)
```
mkdir -p zson && \
for file in zeek-default/*
do
  gzcat "$file" \
      | zq -f zson - \
      | gzip > zson/"$(basename "$file" | sed 's/\.log\.gz//')".zson.gz 
done
```

philrz replied:
RE: Nit1. Hunh, ok... fair point. I guess I personally don't find horizontal scrolling bothersome, particularly when it's reference material that I expect most people to skip over anyway. But now that you say it, I realize I'm probably the oddball in this regard. I've switched to your proposal.

RE: Nit2. Duh. Thanks. I was so focused on shellcheck passing on the main script that I never thought to do it on my README examples. :)

mikesbrown commented on scripts/check_md5sums.sh:
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+set -e
+
+if [ $# -ne 1 ] || ! { [ "$1" = "zson" ] || [ "$1" = "bzson" ]; }; then
```
Nit: Modern bash allows numerical expressions, like `(($# != 1))`. It's a big pain to be sh-compliant, and it doesn't look like we're trying to be. My suggestion would be to use the modern form.

Nit2: I think you can also just use `[[ $1 =~ ^b?zson$ ]]` for another modern touch.

philrz replied:
Thanks for the tips. I've made these changes and they test out fine.

mikesbrown replied:
Sorry to sound like a nitter. It seems like you didn't push the changes discussed above. It sounds like you had decided to, so I wanted to point that out.

philrz replied:
Gah! Thanks for catching that. Stupid mistake. I was working in a scratch directory at one point and I think I made the change there by mistake. Fixed now.

mikesbrown commented on scripts/check_md5sums.sh:
```diff
@@ -0,0 +1,34 @@
+  COMPARE_TO="$(basename "$FILE")"
+  ZPATH=${COMPARE_TO/.${ZNG_TYPE}.gz/}
+  echo -n "${ZPATH}:" | tee -a "$TMPFILE"
+  gzcat zeek-default/"$ZPATH".log.gz | zq -f "$ZNG_TYPE" - | md5sum | awk '{ print $1 }' | tee -a "$TMPFILE"
```
Portability nits: I've found differences in behavior with the convenience *cat and *grep tools on Mac OS vs. GNU/Linux. This works on both, right? And now that I think of it, I don't think Mac OS has `md5sum` by default....

Using `\` would make this more readable also.

philrz replied:
RE: Portability. Thanks for the reminder. I just looked into this and put in a change to use `gzcat` if available (which it is on Mac OS) and `zcat` if not (since that handles `.gz` on Linux, whereas the `zcat` that ships with Mac OS is `.Z` only). Likewise, you were right about `md5sum` so I rigged it up to use `md5 -q` if there's no `md5sum` in the `PATH`. It now tests out fine on both Mac OS and Linux.

And I've put in the `\` you suggested.

Commit:
PR feedback

Commit:
Look for md5 if md5sum not available

philrz commented:
I think I've addressed all PR feedback. If anyone wants to give it another look, feel free. Otherwise I'll just leave this open until the ZNG/BZNG changes come through on the `zq` side at which point I'll do some final renames and then merge this.

mikesbrown commented on scripts/check_md5sums.sh:
```diff
@@ -0,0 +1,56 @@
+#!/bin/bash
+set -e
```
What about `set -eo pipefail`. There's a good long pipe further down and we want to make sure its failure is known.

philrz replied:
Fair point. I've changed it to that and retested and it's fine.

mikesbrown commented on scripts/check_md5sums.sh:
```diff
@@ -0,0 +1,56 @@
+  echo -n "${ZPATH}:" | tee -a "$TMPFILE"
+  "$ZCAT" zeek-default/"$ZPATH".log.gz \
+      | zq -f "$ZNG_TYPE" - \
+      | $SUMTOOL \
```
Does shellcheck complain about this?

philrz replied:
No complaints. In fact, I'd initially put quotes around it precisely because I assumed shellcheck would complain, but then when it actually needs to use the `md5 -q` variant, you get:

```
# ./check_md5sums.sh zson
capture_loss:./check_md5sums.sh: line 40: md5 -q: command not found
```

So this is actually the one time we _want_ it to do the non-quoted expansion thing. Whether shellcheck was smart enough to pick up on that context and that's why it keeps quiet, I cannot say. 🤷‍♂

mikesbrown commented on scripts/check_md5sums.sh:
```diff
@@ -0,0 +1,56 @@
+  echo "  ======> diffs detected! Check for a zq bug or intentional $ZNG_TYPE format change."
+  echo "          Current hashes are in $TMPFILE"
+  exit 1
+fi
```
You don't have to change this, but something a little cleaner is something like:
```
if ! diff "$TMPFILE" md5sums/"$ZNG_TYPE"
then
    # handle error
    exit 1
fi
echo "  ======> No diffs found. $ZNG_TYPE outputs have not changed."
rm -f "$TMPFILE"
```
This means you don't have to deal with `$?` and `$RET`.

philrz replied:
I tried it your way wand it worked fine so I've switched over to that.

Commit:
More PR feedback

Commit:
Final renames to point to zng/bzng now that zq has transitioned

Commit:
Add changed files I missed
